### ISSUE TITLE
for loop snippet ignores var preference

### ISSUE THREAD

**Issue by JamesNK** [dotnet/roslyn#70092]
The snippet for adding a for loop outputs an `int` type even though VS has been configured to prefer `var`.

**Version Used**: 

.NET 8 RC2

**Steps to Reproduce**:

1. Configure VS to prefer var.
2. Type `for` and tab twice to add the snippet.

**Expected Behavior**:

```cs
for (var i = 0; i < length; i++)
{

}
```

**Actual Behavior**:

```cs
for (int i = 0; i < length; i++)
{

}
```

![for-var](https://github.com/dotnet/roslyn/assets/303201/bba9c1bd-a795-4314-9808-78c3b96c16c1)

Simple recording of this issue. `int` is output despite recommending `var`.

**Comment by JamesNK** [comment]
This behavior has annoyed me forever. `dotnet/aspnetcore` uses `var` wherever possible, and the for loop snippet adding `int` is a constant source of inconsistency.

https://github.com/search?q=repo%3Adotnet%2Faspnetcore%20%22for%20(int%20i%22&type=code

**Comment by DoctorKrolic** [comment]
> Type `for` and tab twice to add the snippet

This way you invoke legacy snippet, which is a basic text template and is not smart. In order to use semantic snippet, which should follow your preferences, select the second `for` item in the completion list:
![devenv_D8SGNYC8N1](https://github.com/dotnet/roslyn/assets/70431552/ae9fe995-7b2a-46bf-8763-a3a1077f6d68)

**Comment by CyrusNajmabadi** [comment]
Note: this interaction model should just be a point in time thing.  Existing snippets shoudl be superseded by semantic snippets when both are available.  @akhera99 as i think this is something you were plannign on doing.

**Comment by JamesNK** [comment]
Both options are available today for `for`. Give precedence to semantic snippets when two are available with the same name to make the better option the default?

**Comment by JamesNK** [comment]
I tested out the second for option and got this:

```
for (global::System.Int32 i = 0; i < length; i++)
{
    
}
```
😬

**Comment by CyrusNajmabadi** [comment]
> Both options are available today for `for`. Give precedence to semantic snippets when two are available with the same name to make the better option the default?

As I mentioned, it's a point in time issue. :-).

**Comment by DoctorKrolic** [comment]
> Give precedence to semantic snippets when two are available with the same name to make the better option the default?

Semantic snippets where designed to better align with VS Code's snippets behaviour, thus they are breaking for muscle memory. The most obvious difference is that you activate legacy snippets with double tab (first tab inserts snippet text and second expands it to an actual snippet) while semantic snippets are activated by only 1 tab press. The less obvious difference is that in legacy snippets by pressing tab after expanding the snippet you loop through placeholders and need to press enter to jump to the target position while in semantic snippets pressing tab after they are expanded just jumps through placeholders and end up in the target position without looping. Both this changes make semantic snippets fully muscle-compatable with VS Code's behaviour, but break muscle memory for VS users. The idea is you get a tradeoff between having a more powerful tool (semantic snippets) and your muscle memory (old snippets). As time goes on more and more devs switch to semantic snippets and break their existing habits, so in the long run semantic snippets should definitely become the default option. Those snippets that do not correspond to a C# keyword (e.g. `prop`) are already delivirede as semantic snippets only when semantic snippets are enabled in options menu (and I don't know whether this option is on by default, I just enabled it as soon as this feature dropped).

> I tested out the second for option and got this:

Yes, this is a known bug in the simplifier. In short, if you execute a semantic snippet that requires simplification in local functions, lambdas, anonymous delegates or inside other statements (e.g. `for` nested in an `if` block), simplifier cannot find correct node to simplify and you end up with what you've observed here. This is the second concern when it comes to enabling semantic snippets to be the default completion item. Right now there are several critical bugs like this one that makes working with semantic snippets very painful in some cases. I don't believe we can talk about enabling semantic snippets by default until such issues are resolved. People will just get mad at the fact that we broke their muscle memory and delivered such buggy behavior in return and we will just lose people's trust in the feature. This is definitely not the thing we want as the feature itself is very powerful and has a lot of potential to improve people's productivity (saying that as a very active user of semantic snippets (especially postfix ones) on my own)
